edk2toollib/acpi/ivrs_parser.py

```diff
@@ -0,0 +1,643 @@
+        retval = str(self.ivrs_table)
+
+        for sub in self.ivrs_table.SubStructs:
+            retval += str(sub)
+
+        return retval
+
+    def toBytes(self):
+        bytesarray = bytearray(IVRS_TABLE.IVRSHeaderLength)
+        struct.pack_into(IVRS_TABLE.ACPI_TABLE_HEADER.struct_format,
+                         bytesarray,
+                         0,
+                         self.ivrs_table.Signature,
+                         self.ivrs_table.Length,
+                         self.ivrs_table.Revision,
+                         self.ivrs_table.Checksum,
+                         self.ivrs_table.OEMID,
+                         self.ivrs_table.OEMTableID,
+                         self.ivrs_table.OEMRevision,
+                         self.ivrs_table.CreatorID,
+                         self.ivrs_table.CreatorRevision,
+                         self.ivrs_table.IVinfo,
+                         self.ivrs_table.Reserved)
+        bytesarray += self.data
+        return bytesarray
+
+    @staticmethod
+    def byteSum32(val):
+        if (type(val)) is bytes:
+            if len(val) > 4:
+                raise Exception("Invalid 32 bit value: %d", len(val))
+            sum = 0
+            for ele in val:
+                sum += ele
+        else:
+            if (val & 0xFFFFFFFF00000000) != 0:
+                raise Exception("Invalid 32 bit value: %x", val)
+            else:
+                sum = ((val & 0xFF) + ((val >> 8) & 0xFF) + ((val >> 16) & 0xFF) + ((val >> 24) & 0xFF))
+        return sum
+
+    @staticmethod
+    def byteSum64(val):
+        if (type(val)) is bytes:
+            if len(val) > 8:
+                raise Exception("Invalid 64 bit value: %x", val)
+            sum = 0
+            for ele in val:
+                sum += ele
+        else:
+            sum = (val & 0xFF) + ((val >> 8) & 0xFF) +\
+                  ((val >> 16) & 0xFF) + ((val >> 24) & 0xFF) +\
+                  ((val >> 32) & 0xFF) + ((val >> 40) & 0xFF) +\
+                  ((val >> 48) & 0xFF) + ((val >> 56) & 0xFF)
+        return sum
+
+    @staticmethod
+    def validateChecksum8(data):
+        sum = 0
+        for byte in data:
+            sum += byte
+        return (0x100 - (sum & 0xFF)) & 0xFF
+
+    def addIVHDEntry(self, ivhd):
+        # append raw data, update length and checksum
+        self.data += ivhd.data
+        self.ivrs_table.Length += ivhd.Length
+        self.ivrs_table.SubStructs.append(ivhd)
+        self.ivrs_table.Checksum = self.ivrs_table.calculateACPISum(self.data)
+
+    def addIVMDEntry(self, ivmd):
+        # append raw data, update length and checksum
+        self.data += ivmd.data
+        self.ivrs_table.Length += ivmd.Length
+        self.ivrs_table.IVMDlist.append(ivmd)
+        self.ivrs_table.Checksum = self.ivrs_table.calculateACPISum(self.data)
+
+    def IVRSBitEnabled(self):
+        return bool(self.ivrs_table.IVRSBit)
+
+    def CheckIVMDCount(self, goldenxml=None):
```

Comment:
Is this just a test function? Why is it a first-class member of the object? Is there a use-case other than testing?

Reply:
This is a test function. It will be moved to one layer up at consumer (audit level).